browser/ui/webui/brave_settings_ui.cc

```diff
@@ -239,3 +235,15 @@ void BraveSettingsUI::BindInterface(
       web_ui()->GetWebContents()->GetBrowserContext())
       ->BindInterface(std::move(pending_receiver));
 }
+
+void BraveSettingsUI::BindInterface(
+    mojo::PendingReceiver<ai_chat::mojom::AIChatSettingsHelper>
+        pending_receiver) {
+#if BUILDFLAG(ENABLE_AI_CHAT)
+  auto assistant_handler =
+      std::make_unique<settings::BraveLeoAssistantHandler>();
+  auto* assistant_handler_ptr = assistant_handler.get();
+  web_ui()->AddMessageHandler(std::move(assistant_handler));
+  assistant_handler_ptr->BindInterface(std::move(pending_receiver));
```

Comment:
Can you write here on github or in a comment about why you can't just bind before passing to AddMessageHandler. Seems a bit dicy to hold on to a ptr after moving it somewhere else.

Reply:
if i change the order here, i get the following crash

```
Received signal 11 SEGV_ACCERR 000000000020
0   libbase.dylib                       0x00000001059d0654 base::debug::CollectStackTrace(void const**, unsigned long) + 28
1   libbase.dylib                       0x00000001059b897c base::debug::StackTrace::StackTrace() + 24
2   libbase.dylib                       0x00000001059d052c base::debug::(anonymous namespace)::StackDumpSignalHandler(int, __siginfo*, void*) + 1208
3   libsystem_platform.dylib            0x00000001857e9a24 _sigtramp + 56
4   libchrome_dll.dylib                 0x00000001110401bc mojo::ReceiverSetBase<mojo::Receiver<ai_chat::mojom::AIChatSettingsHelper, mojo::RawPtrImplRefTraits<ai_chat::mojom::AIChatSettingsHelper>>, void>::AddImpl(ai_chat::mojom::AIChatSettingsHelper*, mojo::PendingReceiver<ai_chat::mojom::AIChatSettingsHelper>, mojo::ReceiverSetContextTraits<void>::Empty, scoped_refptr<base::SequencedTaskRunner>, std::__Cr::unique_ptr<mojo::MessageFilter, std::__Cr::default_delete<mojo::MessageFilter>>) + 80
5   libchrome_dll.dylib                 0x00000001110401bc mojo::ReceiverSetBase<mojo::Receiver<ai_chat::mojom::AIChatSettingsHelper, mojo::RawPtrImplRefTraits<ai_chat::mojom::AIChatSettingsHelper>>, void>::AddImpl(ai_chat::mojom::AIChatSettingsHelper*, mojo::PendingReceiver<ai_chat::mojom::AIChatSettingsHelper>, mojo::ReceiverSetContextTraits<void>::Empty, scoped_refptr<base::SequencedTaskRunner>, std::__Cr::unique_ptr<mojo::MessageFilter, std::__Cr::default_delete<mojo::MessageFilter>>) + 80
6   libchrome_dll.dylib                 0x000000011104001c mojo::ReceiverSetBase<mojo::Receiver<ai_chat::mojom::AIChatSettingsHelper, mojo::RawPtrImplRefTraits<ai_chat::mojom::AIChatSettingsHelper>>, void>::Add(ai_chat::mojom::AIChatSettingsHelper*, mojo::PendingReceiver<ai_chat::mojom::AIChatSettingsHelper>, scoped_refptr<base::SequencedTaskRunner>) requires !internal::kIsRuntimeFeatureGuarded<mojo::ReceiverSetTraits<T>::InterfaceType> + 72
7   libchrome_dll.dylib                 0x000000011103ffbc ai_chat::AIChatSettingsHelper::BindInterface(mojo::PendingReceiver<ai_chat::mojom::AIChatSettingsHelper>) + 52
8   libchrome_dll.dylib                 0x00000001107b5c24 settings::BraveLeoAssistantHandler::BindInterface(mojo::PendingReceiver<ai_chat::mojom::AIChatSettingsHelper>) + 40
9   libchrome_dll.dylib                 0x000000011078f4a8 BraveSettingsUI::BindInterface(mojo::PendingReceiver<ai_chat::mojom::AIChatSettingsHelper>) + 68
10  libchrome_dll.dylib                 0x000000011103aabc bool content::internal::SafeDownCastAndBindInterface<ai_chat::mojom::AIChatSettingsHelper, BraveSettingsUI>(content::RenderFrameHost*, mojo::PendingReceiver<ai_chat::mojom::AIChatSettingsHelper>&) + 136

```

Reply:
added comment

Reply:
Yes, using storing raw_ptr from unique_ptr is not that bad. But I would suggest `AIChatSettingsHelper` be created here and passed to the BraveLeoAssistantHandler's constructor if lifetime is what force us to have code as is. Also that would make BraveLeoAssistantHandler::RegisterMessages() single job, by removing code for creating settings helper.

```
  auto assistant_handler =
      std::make_unique<settings::BraveLeoAssistantHandler>(
          std::make_unique<ai_chat::AIChatSettingsHelper>(
              web_ui()->GetWebContents()->GetBrowserContext());
```

Reply:
ah, that looks neat. changed.